031-statefulset-removal.md

```diff
@@ -0,0 +1,225 @@
+Therefore it was decided to investigate what options we have to replace StatefulSets to gain more flexibility and be able to continue the Strimzi development.
+Below are some of the features and use cases which might be significantly easier to implement without the StatefulSet limitations.
+
+*Note: This proposal does not cover these features and use-case.
+They are just used to explain the things which might be easier when this proposal is implemented.*
+
+### Individual broker configuration
+
+Being able to individually configure the brokers would allow us to improve Strimzi in several areas:
+* Less rolling updates
+* Asymmetric configurations of storage or CPU / memory resources
+
+### ZooKeeper-less Kafka
+
+The removal of ZooKeeper from Kafka will bring a lot of advantages but also some challenges.
+There will no more be a clear separation between the ZooKeeper cluster and the Kafka cluster.
+Any Kafka broker can be in theory configured as controller.
+With a single StatefulSet managing all Kafka pods, it would be hard to handle all the possible configurations:
+* Dedicated set of controller nodes
+* All nodes being both controllers and brokers
+* Some nodes being both brokers and controller and some only brokers
+
+We could use multiple StatefulSets to separate the different types of nodes.
+But even then it would be hard to change the cluster topologies and the node types.
+Without the StatefulSet limitations, we should have a lot more flexibility to mix the node types and their roles.
+
+### Stretch clusters
+
+Pods managed by a single StatefulSet always exist only on single Kubernetes cluster.
+If you want to have a Kafka cluster which would stretch multiple Kubernetes clusters, you would need to create a StatefulSet on each of them and have it create the Pods on given cluster.
+But you would not be able easily move the brokers between the clusters since they would depend on the StatefulSet index numbers to generate their broker IDs.
+
+## Proposal
+
+The StatefulSets should be replaced with a new Strimzi custom resource called `KafkaNodeSet`.
+KafkaNodeSet will be used only internally.
+Users will not be expected to interact with.
+
+*Note: The name KafkaNodeSet was chosen to be in sync with our other resources.
+But the same KafkaNodeSet custom resource definition will be used for ZooKeeper as well as for Kafka.
+There will be no ZooKeeperNodeSet.*
+
+The KafkaNodeSets will have a following structure:
+* The usual metadata, kind and API identifications
+* A `.spec` section with
+    * Pod selector to be able to find the pods managed by this KafkaNodeSet
+    * List of Pods (full Pod YAML definition including metadata section, but without status section)
+* A status section with
+    * Number of managed, current and ready pods
+    * List of error conditions in case of problems
+    * Observed generation
+
+The node sets will be used very similarly to the StatefulSets.
```

Comment:
Probably worth mentioning:

* that the Pods will be in the same namespace as the NodeSet (assuming that's true)?
* that the NodeSet would be the owner of the Pods and GC will be used